src/v/cloud_topics/dl_stm/dl_stm_state.cc

```diff
@@ -48,6 +46,8 @@ void dl_stm_state::push_overlay(dl_version version, dl_overlay overlay) {
       // The overlay becomes visible starting with the current version.
       .added_at = version,
     });
+
+    _version_invariant.set_version(version);
```

Comment:
I don't recall the logic behind this so I'm having a hard time to review without digging into it but can you write a test that shows the need for set_version to happen here? Surprised this doesn't break any test. Also surprised this is needed because the vassert is `version >= _last_version && version > _last_snapshot_version` 🤔